### PR TITLE
feat: add `useBackgroundQuery` and `useReadQuery` Suspense hooks

### PR DESCRIPTION
**Pull request by alessbell**
Adds `useBackgroundQuery` and `useReadQuery` hooks as companion hooks to `useSuspenseQuery` for fetching and reading data within React Suspense boundaries.

## `useBackgroundQuery`

`useBackgroundQuery` allows a user to initiate a query as "high up" in their application's render tree as possible, decoupled from the site where the returned data is consumed, which will trigger the nearest Suspense boundary while the query is pending.

The hook accepts `query` and `options` as positional arguments, the latter extending `SuspenseQueryHookOptions` while omitting `returnPartialData` and `refetchWritePolicy` for a slightly smaller initial API surface. The hook returns a `queryRef` as well as `fetchMore` and `refetch` functions.

### `queryRef`

In order to link a query initiated by a specific `useBackgroundQuery` call to the place its data are consumed—which can be uniquely identified not only by the query and variables passed, but also the optional `queryKey` supplied by the user—the hook needs to return a reference to the query that can be read later on.

A `queryRef` is a `QuerySubscription` instance that holds references to the promises that are read or thrown by React's `use` (or Apollo Client's `__use` if `React.use` is not detected).

## `useReadQuery`

`useReadQuery` accepts a `queryRef` and returns `data`, `networkStatus` and `error`.

> `networkStatus` is only useful in certain edge cases since the vast majority of the time the component is in a suspended state when going to the network via loading/refetching. One of these cases is captured in the test `does not suspend deferred queries with data in the cache and using a "cache-and-network" fetch policy`: the `networkStatus` on initial render is `loading` since the component does not suspend when the entire query is already available in the cache.

## Hello World Example

```tsx
interface QueryData {
  foo: { bar: string };
}

const suspenseCache = new SuspenseCache();

const client = new ApolloClient({
  uri: "http://localhost:4000/graphql",
  cache: new InMemoryCache()
});

function SuspenseFallback() {
  return <div>Loading...</div>;
}

function Child({ queryRef }: { queryRef: QuerySubscription<QueryData> }) {
  const { data } = useReadQuery<QueryData>(queryRef);

  return <div>{data.foo.bar}</div>;
}

function Parent() {
  const [queryRef] = useBackgroundQuery(query);

  return <Child queryRef={queryRef} />;
}

function App() {
  return (
    <ApolloProvider client={client} suspenseCache={suspenseCache}>
      <ErrorBoundary fallback={<div>Error</div>}>
        <Suspense fallback={<SuspenseFallback />}>
          <Parent />
        </Suspense>
      </ErrorBoundary>
    </ApolloProvider>
  );
}
```

### Future PRs
- type-related:
  - improve type signature of `refetch`
  - test/refine type inference of returned `data` (this handoff happens across the two hooks, we may need to update `QuerySubscription`)
- additional test coverage: 
  - error policy tests: error policy is a supported option, and I've added some tests to this effect in this PR, but additional test coverage for this option will come in a follow-up PR
  - `queryKey` tests
- compatibility with Next.js package https://github.com/apollographql/apollo-client/issues/10875
- docs x2 https://github.com/apollographql/apollo-client/issues/10874

### CODE REVIEW

File: src/react/hooks/__tests__/useSuspenseQuery.test.tsx [hunk 1]
```diff
@@ -163,8 +163,8 @@ function useSimpleQueryCase() {
 function usePaginatedCase() {
   interface QueryData {
     letters: {
-      name: string;
```
Comment (alessbell) [inline src/react/hooks/__tests__/useSuspenseQuery.test.tsx:166]: Noticed this when re-using this test structure for `useBackgroundQuery`.

File: src/react/hooks/useBackgroundQuery.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+export function useBackgroundQuery_experimental<
+  TData = any,
+  TVariables extends OperationVariables = OperationVariables
+>(
+  query: DocumentNode | TypedDocumentNode<TData, TVariables>,
+  options: Omit<
+    SuspenseQueryHookOptions<TData, TVariables>,
+    'returnPartialData' | 'refetchWritePolicy'
```
Comment (alessbell) [inline src/react/hooks/useBackgroundQuery.ts:41]: The two unsupported-at-first options from `SuspenseQueryHookOptions` — I haven't had a chance to test these yet, and it seemed like a reasonable place to draw the line for the initial PR.

File: src/react/hooks/__tests__/useBackgroundQuery.test.tsx [hunk 3]
```diff
@@ -0,0 +1,2290 @@
...
+      expect(_result.data.results.length).toBe(6);
+      expect(resultSet.size).toBe(6);
+      expect(values).toEqual([0, 1, 1, 2, 3, 5]);
+    });
+  });
+
+  describe('fetch policy behaviors', () => {
+    describe.skip('cache-and-network', () => {
+      // TODO: should return cache data first if it exists
+      it('returns initial cache data followed by network data', async () => {
```
Comment (jerelmiller) [inline src/react/hooks/__tests__/useBackgroundQuery.test.tsx:901]: As mentioned before, I generally prefer tests to be as [flat](https://kentcdodds.com/blog/avoid-nesting-when-youre-testing) as possible. 

One of the biggest reasons for this is that as I'm reading a test and understanding what we are looking at, I use the test description as a means to understand what the test is meant for. In a lot of tests where I start to see nested `describes`, the description starts to get split between the `describe` and the `it`/`test` description. This gets problematic the more the file grows since it now requires me to scroll to understand the whole context for the test.

Instead, I prefer more descriptive tests with a flat structure. This allows me to stay isolated to the test without the need to scroll around to find additional context.

For this test, that would include adding the `cache-and-network` qualifier in the test description itself.

```suggestion
      it('returns initial cache data followed by network data when the fetch policy is `cache-and-network`', async () => {
```

I know its longer and some of it is repetitive, but again, I can look at this test and not have to scroll around to find additional context.

Thoughts?

File: src/react/hooks/__tests__/useBackgroundQuery.test.tsx [hunk 4]
```diff
@@ -0,0 +1,2290 @@
...
+      const values = Array.from(resultSet).map((item) => item.value);
+
+      expect(_result.data).toEqual({ results });
+      expect(_result.data.results.length).toBe(6);
+      expect(resultSet.size).toBe(6);
+      expect(values).toEqual([0, 1, 1, 2, 3, 5]);
+    });
+  });
+
+  describe('fetch policy behaviors', () => {
```
Comment (jerelmiller) [inline src/react/hooks/__tests__/useBackgroundQuery.test.tsx:898]: [nit] While `describe` is billed as a grouping mechanism, I'd challenge the usefulness of this block here. This doesn't give too much additional context to the tests themselves since the tests do a pretty decent job of describing what they are testing. If possible, I'd prefer if this were removed entirely and just allow the tests inside of here to be a bit flatter (which also means less indentation!)

 tend to find this unnecessary as it doesn't really do much for my test. As this file grows, this gets farther and farther away from the tests grouped inside of here

**Review by jerelmiller — CHANGES_REQUESTED** [review]
I had a lot of [nit] comments that you can take or leave. I'm requesting changes to see if we can update the exports to only export public APIs before this is merged. Once that is done, I'd be happy to re-reviewing and approve. 

Really great work here! Excited to get an initial release on this! 🎉 🎉 🔥 🔥

File: src/__tests__/__snapshots__/exports.ts.snap [hunk 5]
```diff
@@ -53,15 +53,21 @@ Array [
...
   "useFragment_experimental",
   "useLazyQuery",
   "useMutation",
+  "usePromiseVersion",
   "useQuery",
   "useReactiveVar",
+  "useReadQuery_experimental",
   "useSubscription",
   "useSuspenseQuery_experimental",
+  "useTrackedSubscriptions",
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/__tests__/__snapshots__/exports.ts.snap]: Yep sounds good! Updated in https://github.com/apollographql/apollo-client/pull/10755/commits/7cd2303141b022e6afb6360e29d9e95f9b21fe0c.

File: src/react/hooks/__tests__/useBackgroundQuery.test.tsx (hunk 4 above)
Comment (alessbell) [reply src/react/hooks/__tests__/useBackgroundQuery.test.tsx]: One thing I like about `describe` blocks is the ability to call `.skip`/`.only` on groups of related tests which helps avoid jumping around a long file like this more than necessary. That said, once the need to rapidly iterate is over I'm definitely not attached to these groupings :) Removed in [e473774](https://github.com/apollographql/apollo-client/pull/10755/commits/e473774a74f45db8d6c7a93a197af333fec7abb0).

File: src/react/hooks/__tests__/useBackgroundQuery.test.tsx [hunk 6]
```diff
@@ -0,0 +1,2290 @@
...
+    const _result = await queryRef.promises.main;
+
+    expect(_result).toEqual({
+      data: { hello: 'world 1' },
+      loading: false,
+      networkStatus: 7,
+    });
+  });
+
+  describe('hook options', () => {
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/hooks/__tests__/useBackgroundQuery.test.tsx]: [e473774](https://github.com/apollographql/apollo-client/pull/10755/commits/e473774a74f45db8d6c7a93a197af333fec7abb0)

File: src/react/hooks/__tests__/useBackgroundQuery.test.tsx [hunk 7]
```diff
@@ -0,0 +1,2290 @@
...
+    const suspenseCache = new SuspenseCache();
+    const mocks = [
+      {
+        request: { query },
+        result: { data: { hello: 'world 1' } },
+      },
+    ];
+    const { result } = renderHook(() => useBackgroundQuery(query), {
+      wrapper: ({ children }) => (
+        <MockedProvider mocks={mocks} suspenseCache={suspenseCache}>
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/hooks/__tests__/useBackgroundQuery.test.tsx]: Thanks for that reminder, updated in [e473774](https://github.com/apollographql/apollo-client/pull/10755/commits/e473774a74f45db8d6c7a93a197af333fec7abb0) 👍

File: src/react/hooks/__tests__/useBackgroundQuery.test.tsx [hunk 8]
```diff
@@ -0,0 +1,2290 @@
...
+          onClick={() => {
+            const fetchMoreOpts: FetchMoreQueryOptions<Variables, QueryData> & {
+              updateQuery?: (
+                previousQueryResult: QueryData,
+                options: {
+                  fetchMoreResult: QueryData;
+                  variables: Variables;
+                }
+              ) => QueryData;
+            } = { variables: { offset: 0, limit: 2 } };
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/hooks/__tests__/useBackgroundQuery.test.tsx]: Good catch, fixed!

File: src/react/hooks/index.ts [hunk 9]
```diff
@@ -8,3 +8,4 @@ export * from './useSubscription';
 export * from './useReactiveVar';
 export * from './useFragment';
 export * from './useSuspenseQuery';
+export * from './useBackgroundQuery';
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/hooks/index.ts]: Totally agree. Updated in https://github.com/apollographql/apollo-client/pull/10755/commits/7cd2303141b022e6afb6360e29d9e95f9b21fe0c.

File: src/react/cache/QueryReference.ts [hunk 10]
```diff
@@ -66,7 +67,7 @@ export class QuerySubscription<TData = unknown> {
       this.refetching = false;
     }
 
-    this.subscription = observable.subscribe({
+    this.queryRef = observable.subscribe({
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/cache/QueryReference.ts]: Ah yep good call, I see I went a bit too far here :) Reverted as part of https://github.com/apollographql/apollo-client/pull/10755/commits/e473774a74f45db8d6c7a93a197af333fec7abb0.

File: src/react/hooks/__tests__/useBackgroundQuery.test.tsx (hunk 3 above)
Comment (alessbell) [reply src/react/hooks/__tests__/useBackgroundQuery.test.tsx]: Updated test name in https://github.com/apollographql/apollo-client/pull/10755/commits/5b1d3ee60a81ad9b361ef79eb857d3079c234144

File: src/react/hooks/__tests__/useBackgroundQuery.test.tsx [hunk 11]
```diff
@@ -0,0 +1,2290 @@
...
+  }
+
+  function Child({
+    queryRef,
+    fetchMore,
+  }: {
+    fetchMore: FetchMoreFunction<QueryData, OperationVariables>;
+    queryRef: QueryReference<QueryData>;
+  }) {
+    const { data, error } = useReadQuery<QueryData>(queryRef);
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/hooks/__tests__/useBackgroundQuery.test.tsx]: Yes thanks for the reminder here! Removed a bunch of unneeded generics args passed to hooks in useBackgroundQuery and useSuspenseQuery tests in this commit: https://github.com/apollographql/apollo-client/pull/10755/commits/5b1d3ee60a81ad9b361ef79eb857d3079c234144

File: src/react/hooks/__tests__/useBackgroundQuery.test.tsx [hunk 12]
```diff
@@ -0,0 +1,2290 @@
...
+    foo: { bar: string };
+  }
+
+  function SuspenseFallback() {
+    renders.suspenseCount++;
+    return <div>loading</div>;
+  }
+
+  function Child({ queryRef }: { queryRef: QueryReference<QueryData> }) {
+    const { data } = useReadQuery<QueryData>(queryRef);
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/hooks/__tests__/useBackgroundQuery.test.tsx]: Resolved in https://github.com/apollographql/apollo-client/pull/10755/commits/5b1d3ee60a81ad9b361ef79eb857d3079c234144

File: src/react/hooks/__tests__/useBackgroundQuery.test.tsx [hunk 13]
```diff
@@ -0,0 +1,2290 @@
...
+
+  function Parent() {
+    const { queryRef } = useBackgroundQuery(query);
+    // count renders in the parent component
+    renders.count++;
+    return <Child queryRef={queryRef} />;
+  }
+
+  function ParentWithVariables({
+    variables,
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/hooks/__tests__/useBackgroundQuery.test.tsx]: [397727e](https://github.com/apollographql/apollo-client/pull/10755/commits/397727e792994c5b434259d64848d4fa2514d2b1)

File: src/react/hooks/__tests__/useBackgroundQuery.test.tsx [hunk 14]
```diff
@@ -0,0 +1,2290 @@
...
+    queryRef: QueryReference<QueryData>;
+  }) {
+    const { data, error } = useReadQuery<QueryData>(queryRef);
+
+    return (
+      <div>
+        {error ? <div>{error.message}</div> : null}
+        <button
+          onClick={() => {
+            const fetchMoreOpts: FetchMoreQueryOptions<Variables, QueryData> & {
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/hooks/__tests__/useBackgroundQuery.test.tsx:436]: Hmmm, this is a tricky one. I don't think passing any of the functions vended by these hooks as props—especially in the case of `useBackgroundQuery` + `useReadQuery`, where they'll almost certainly be called in a child component—should be considered an anti-pattern. These functions (I'm thinking of `fetchMore` and `refetch` specifically here) shouldn't need to be wrapped in an enclosing function—which itself you'd almost certainly want to wrap in `useCallback`, something easy to forget (absent above), and which our vended functions already do—to be used ergonomically.

IMO the fact that the types are cumbersome here is something to solve for (note I added a line to the future PRs section to make the `fetchMore` type more user-friendly here; `refetch` don't suffer this problem and passing `refetch` without wrapping it was very ergonomic in another test.)

So, I think I'll leave this test as-is, and remove the cumbersome types when they become unnecessary in the future :)

File: src/react/hooks/useBackgroundQuery.ts [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+      const promise = queryRef.refetch(variables);
+      setVersion('network');
+      return promise;
+    },
+    [queryRef]
+  );
+
+  queryRef.version = version;
+
+  return useMemo(() => {
```
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/hooks/useBackgroundQuery.ts:85]: +1! I will be making this change as part of this PR. It's the last thing on my list :)

**Comment by phryneas** [comment]
One thing to keep in mind: we'll have to try this with the Next package, to make sure that those `queryRef` objects work nicely there, too. I think they *should*, but let's be sure.

**Comment by alessbell** [comment]
> One thing to keep in mind: we'll have to try this with the Next package, to make sure that those queryRef objects work nicely there, too. I think they should, but let's be sure.

Good call-out, added to the list of future work on the PR description that I'll tackle... next 😀

**Comment by phryneas** [comment]
Looking at this, it might sense for us to pair on this.

Quick intro: take a look at [these wrapped hook definitions](https://github.com/apollographql/apollo-client-nextjs/blob/main/package/src/ssr/hooks.ts).
In the Next.js wrapper, we wrap the hook into a version that transports a partial copy of the result from server to browser, to make sure that the browser can render exactly the same HTML with the same data that the server had - at least on first render.  
So, looking at the `queryReference` class, we would need to transport over most of `QueryReference.result`, analogous to `useQuery`.  
But there are more values that we should transport over, that might not be possible with the simple wrapping mechanism that we currently use. 

Going through the public properties, `observable` is probably fine, just as `version` - but we need to take a good look at what we do with `promises` - do we transport the resolved value of those promises over, and recreate promises on the client side?  
What if they already had passed through `use`, do we need to mock the properties added by `use` as well?  
Note that these problems might go away if we decided that `version` and `promise` could be `private` instead of public, as we then know that no user would just play around with them - and we could manage most of that by just wrapping `useReadQuery`, while ignoring those problematic properties of `QueryReference`.

File: src/react/hooks/useBackgroundQuery.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (alessbell) [reply src/react/hooks/useBackgroundQuery.ts:86]: Resolved in [f4393f8](https://github.com/apollographql/apollo-client/pull/10755/commits/f4393f84ba5450407c4e669b83b3570600b7fcd7)

File: src/react/hooks/__tests__/useBackgroundQuery.test.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (jerelmiller) [reply src/react/hooks/__tests__/useBackgroundQuery.test.tsx:438]: Ya thats a good point. This is probably one of those "style differences in different developers" situations 😄. Given the amount of overhead to type this though, this might be a signal that we need to more closely look at this since this is what a user would need to do to type this properly. 

Not something we need to address in this PR! Thanks for the dialogue here!

File: src/react/hooks/__tests__/useSuspenseQuery.test.tsx [hunk 16]
```diff
@@ -6889,10 +6889,7 @@ describe('useSuspenseQuery', () => {
       expectTypeOf(inferred).toEqualTypeOf<VariablesCaseData>();
       expectTypeOf(inferred).not.toEqualTypeOf<VariablesCaseData | undefined>();
 
-      const { data: explicit } = useSuspenseQuery<
-        VariablesCaseData,
-        VariablesCaseVariables
-      >(query);
+      const { data: explicit } = useSuspenseQuery(query);
```
Comment (jerelmiller) [inline src/react/hooks/__tests__/useSuspenseQuery.test.tsx:6892]: These being type tests, the explicit setting of the generic type here was actually purposeful (hence the name `explicit`). Essentially I wanted to make sure that setting the type of the generic directly rather than relying on type inference ensured that `data` was properly typed with the various options you can pass to the hook.

I realize thought that I made a mistake and I shouldn't be passing a query here that uses `TypedDocumentNode` (which make type inference work correctly). I'll update this in a separate PR to use an untyped query document to test against, which is a bit more realistic.

In the mean time, would you mind reverting these `useSuspenseQuery` type tests back to the original?

Comment (alessbell) [reply src/react/hooks/__tests__/useSuspenseQuery.test.tsx:6892]: Oops, my bad - was moving too fast there. Totally unintended, will revert the changes for the types tests now.

Comment (alessbell) [reply src/react/hooks/__tests__/useSuspenseQuery.test.tsx]: [56083fb](https://github.com/apollographql/apollo-client/pull/10755/commits/56083fb23cfe38ce1f89e3f61bc755e577c6098f) :)

**Review by jerelmiller — APPROVED** [review]
Outside of my last comment, this PR looks great! I'm super excited to get this into an alpha!
